ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/TimeServiceIT.scala

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.api.testtool.suites
+
+import com.daml.error.definitions.LedgerApiErrors
+import com.daml.ledger.api.testtool.infrastructure.Allocation._
+import com.daml.ledger.api.testtool.infrastructure.Assertions._
+import com.daml.ledger.api.testtool.infrastructure.LedgerTestSuite
+import com.daml.ledger.client.binding.{Primitive => P}
+import com.daml.ledger.test.semantic.TimeTests._
+import io.grpc.Status
+import java.time.Instant
+
+import scala.concurrent.Future
+
+final class TimeServiceIT extends LedgerTestSuite {
+
+  test(
+    "TSTimeIsStatic",
+    "Time stands still when static time enabled",
+    allocate(NoParties),
+    runConcurrently = false,
+    enabled = _.staticTime,
+    disabledReason = "requires user static time feature",
+  )(implicit ec => { case Participants(Participant(ledger)) =>
+    for {
+      initialTime <- ledger.time()
+      _ <- Future { Thread.sleep(100) }
+      laterTime <- ledger.time()
+    } yield {
+      assertEquals("ledger time should stand still", laterTime, initialTime)
+    }
+  })
+
+  test(
+    "TSTimeCanBeAdvanced",
+    "Time can be advanced when static time enabled",
+    allocate(NoParties),
+    runConcurrently = false,
+    enabled = _.staticTime,
+    disabledReason = "requires user static time feature",
+  )(implicit ec => { case Participants(Participant(ledger)) =>
+    for {
+      initialTime <- ledger.time()
+      thirtySecLater = initialTime.plusSeconds(30)
+      _ <- ledger.setTime(initialTime, thirtySecLater)
+      laterTime <- ledger.time()
+    } yield {
+      assertEquals("ledger time should stand still", laterTime, thirtySecLater)
+    }
+  })
+
+  test(
+    "TSFailWhenTimeNotAdvanced",
+    "The submission of an exercise before time advancement should fail",
+    allocate(SingleParty),
+    runConcurrently = false,
+    enabled = _.staticTime,
+    disabledReason = "requires user static time feature",
+  )(implicit ec => { case Participants(Participant(ledger, party)) =>
+    for {
+      initialTime <- ledger.time()
+      thirtySecLater <- createTimestamp(initialTime.plusSeconds(30))
+      checker <- ledger.create(party, TimeChecker(party, thirtySecLater))
+      failure <- ledger
+        .exercise(party, checker.exerciseTimeChecker_CheckTime(_))
+        .mustFail("submitting choice prematurely")
+    } yield {
+      assertGrpcError(
+        ledger,
+        failure,
+        Status.Code.INVALID_ARGUMENT,
+        LedgerApiErrors.CommandExecution.Interpreter.GenericInterpretationError,
+        Some("Unhandled exception"),
```

Comment:
this doesn't look good

Reply:
This is the text of the DAML level exception and the consequence of the fact that the assert is not handled in the daml code. It was a deliberate choice on my part not to catch it because I want a transaction failure as an observable effect:
```
Invalid argument: Command interpretation error in LF-DAMLe: Interpretation error: Error: Unhandled exception: DA.Exception.AssertionFailed:AssertionFailed@3f4deaf1{ message = "Assertion failed" }. Details: Last location: [DA.Internal.Exception:168], partial transaction: root node NodeId(0): Exercise(ContractId(000877ea407449057ee5eb96f357e7670028d2b85339d3b81c5d3d1ad3ba830702),1e0da60554f5dda79bbf8924da083a2cea3c2b65c6194ad89d279d1ec90436c5:TimeTests:TimeChecker,TimeChecker_CheckTime,false,TreeSet(TSFailWhenTimeNotAdvanced-alpha-1a5f888b2f2bb-party-0),ValueRecord(None,ImmArray()),TreeSet(TSFailWhenTimeNotAdvanced-alpha-1a5f888b2f2bb-party-0),TreeSet(TSFailWhenTimeNotAdvanced-alpha-1a5f888b2f2bb-party-0),TreeSet(),ImmArray(),None,None,false,None,V14).
```

Reply:
Your logic handles what it gets. What I think could be improved is the content of the error: the error says `Unhandled exception`, doesn't clearly tell the user what happened, but somehow it is a `FAILED_PRECONDITION` gRPC error code.
By checking the `RejectionGenerators` code, I see that `com.daml.lf.interpretation.Error.UnhandledException` is mapped to this one which might be wrong (unhandled should mean `INTERNAL`). On the other hand, at the same time it's not sound to me to be able to generate this LF `UnhandledException` on expected work-flows.
wdyt ? CC @moritzkiefer-da @remyhaemmerle-da

Reply:
these are exceptions in Daml code users have written, e.g., calls to `error`, failed `assert`s, … not exceptions in our code. From a ledger implementors pov those are not internal errors.

Reply:
Oh, I see. I'm not accustomed to them. Maybe it would be more clear if they say `Unhandled Daml exception` or something like that. But this we could adapt in the error codes API as well

Reply:
Daml-Lf exception might be more precise but good point either way, we’ll take a look if we can change the text. I’ll leave the error codes to you but if you need help for that from our side please reach out.